pkg/server/commands/listusers/list_users_rpc_test.go

```diff
@@ -2567,3 +2569,266 @@ func TestListUsersReadFails_NoLeaks_TTU(t *testing.T) {
+		},
+		{
+			name:        "ttu_no_access",
+			relation:    "ttu",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     1,
+		},
+		{
+			name:        "userset_no_access_1",
+			relation:    "userset",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     1,
+		},
+		{
+			name:        "userset_no_access_2",
+			relation:    "userset",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     1,
+		},
+		{
+			name:        "userset_access",
+			relation:    "userset",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     2,
+		},
+		{
+			name:        "multiple_userset_no_access",
+			relation:    "multiple_userset",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     3,
+		},
+		{
+			name:        "multiple_userset_access",
+			relation:    "multiple_userset",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     2,
+		},
+		{
+			name:        "wildcard_no_access",
+			relation:    "wildcard",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     1,
+		},
+		{
+			name:        "wildcard_access",
+			relation:    "wildcard",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     1,
+		},
+
+		{
+			name:        "union_and_ttu",
+			relation:    "union_and_ttu",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     5,
+		},
+		{
+			name:        "union_and_ttu_no_access",
+			relation:    "union_and_ttu",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     4,
+		},
+		{
+			name:        "union_or_ttu",
+			relation:    "union_or_ttu",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     3,
+		},
+		{
+			name:        "union_or_ttu_no_access",
+			relation:    "union_or_ttu",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     6,
+		},
+		{
+			name:        "intersection_of_ttus",
+			relation:    "intersection_of_ttus",
+			object:      openfgav1.Object{Type: "document", Id: "1"},
+			userFilters: []*openfgav1.UserTypeFilter{{Type: "user"}},
+			dbReads:     8,
+		},
+	}
+
+	l := NewListUsersQuery(ds)
```

Comment:
This line needs to be before line 2822, otherwise you get a data race on this line: `l.ds = storagewrappers.NewCombinedTupleReader(l.ds, req.GetContextualTuples())`